c10/core/TensorImpl.h

```diff
@@ -907,30 +843,87 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     return std::move(autograd_meta_);
   }
 
-  // NOTE: `shallow_copy_and_detach()` does not copy the following TensorImpl fields:
+  // NOTE [ TensorImpl Shallow-Copying ]
+  //
+  // TensorImpl shallow-copying is used when we want to have two Variables share the same storage pointer
+  // and tensor metadata, but each with a different autograd history. Example call sites:
+  //
+  // 1. `var_detached = var.detach()` uses `shallow_copy_and_detach()` to create `var_detached` that shares
+  // the same storage pointer and tensor metadata with `var`, but with a completely new autograd history.
+  // 2. `var.set_data(tensor)` uses `shallow_copy_from()` to copy storage pointer and tensor metadata from
+  // `tensor` into `var`, while keeping `var`'s original AutogradMeta.
+  //
+  // Functions that shallow-copy a TensorImpl (such as `shallow_copy_and_detach()` / `shallow_copy_from()` /
+  // `copy_tensor_data()`) copy the storage pointer and the tensor metadata fields (e.g. sizes / strides /
+  // storage_offset) by value. However, the following fields are not copied:
+  //
   // 1. the AutogradMeta pointer, because it is unique for each Variable.
-  // 2. the version counter, because it is set to the passed in `version_counter`.
-  //    See NOTE [ Version Counter Sharing ] for details.
+  // 2. the version counter, because the destination TensorImpl's version counter is either set to the
+  // passed-in `version_counter` (in `shallow_copy_and_detach()` and `copy_tensor_data()`), or it is kept
+  // intact (in `shallow_copy_from()`). See NOTE [ Version Counter Sharing ] for details.
   //
-  // NOTE: `allow_tensor_metadata_change` determines whether the TensorImpl shallow-copy
-  // allows changes to its metadata (e.g. sizes / strides / storage / storage_offset).
-  // See NOTE [ Metadata Change for a Detached Tensor ] for details.
+  // In `shallow_copy_and_detach()` and `copy_tensor_data()`, the passed-in `allow_tensor_metadata_change`
+  // determines whether the TensorImpl shallow-copy allows changes to its metadata (e.g. sizes / strides /
+  // storage / storage_offset). See NOTE [ Metadata Change for a Detached Tensor ] for details.
+
+  /**
+   * Copy the storage pointer and the tensor metadata fields (e.g. sizes / strides / storage_offset)
+   * from one TensorImpl to another TensorImpl.
+   *
+   * For usage of `version_counter` and `allow_tensor_metadata_change`, see NOTE [ TensorImpl Shallow-Copying ].
+   */
+  virtual void copy_tensor_data(
```

Comment:
a few issues with this function:
1) this is just used by the public API functions `shallow_copy_and_detach`, `shallow_copy_from`, right?  Then it should be private.
2) This doesn't need to be virtual; actually notice you don't ever use this, so this doesn't even need to be a non-static method.
3) because it's a static method and you aren't using virtual dispatch, you can use the correct static types of inputs.  So you don't have to do scary static_casts in the derived types.

Reply:
err, you probably need to make it protected so the derived types can call it.

Reply:
Fixed.